File: interfaces/src/main/java/org/codroid/interfaces/env/Property.java
/*
 *     Copyright (c) 2021 Zachary. All rights reserved.
 *
 *     This file is part of Codroid.
 *
 *     Codroid is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     Codroid is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with Codroid.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.codroid.interfaces.env;

import org.codroid.interfaces.utils.PathUtils;
import org.codroid.interfaces.utils.TomlKt;

import java.nio.file.Path;
import java.util.Map;

import cc.ekblad.toml.TomlMapper;

public class Property extends Resource {

    protected Map<String, Object> map;
    protected static final TomlMapper mDefaultMapper = TomlKt.defaultTomlMapper();

    public Property(Path path) {
        super(path);
        if (path != null) {
            this.map = TomlKt.decode2Map(path, getTomlMapper());
        }
    }

    public <T> T getAs(Class<T> clazz) {
        return TomlKt.toObject(getAttributes(), clazz);
    }

    public Map<String, Object> getAttributes() {
        return map;
    }

    public Property(AddonEnv addonEnv, String relativePathStr) {
        super(addonEnv, relativePathStr);
        map = TomlKt.decode2Map(toPath(), getTomlMapper());
    }

    public TomlMapper getTomlMapper() {
        return mDefaultMapper;
    }
}

File: interfaces/src/main/java/org/codroid/interfaces/env/Property.kt
package org.codroid.interfaces.env

import cc.ekblad.toml.TomlMapper
import cc.ekblad.toml.decode
import cc.ekblad.toml.model.TomlValue
import cc.ekblad.toml.serialization.from
import cc.ekblad.toml.util.InternalAPI
import org.codroid.interfaces.utils.PathUtils
import org.codroid.interfaces.utils.defaultTomlMapper
import java.io.InputStream
import java.nio.file.Path
import kotlin.jvm.internal.Reflection
import kotlin.reflect.KClass
import kotlin.reflect.KType
import kotlin.reflect.full.createType

@OptIn(InternalAPI::class)
open class Property<T> : Resource {
    var entity: T? = null

    constructor(path: Path?, kType: KType) : super(path) {
        if (path != null) {
            entity = this.getTomlMapper().decode(kType, TomlValue.from(path))
        }
    }

    constructor(addonEnv: AddonEnv, relativePath: String, kType: KType) : super(
        addonEnv,
        relativePath
    ) {
        entity = this.getTomlMapper().decode(kType, TomlValue.from(this.toPath()))
    }

    constructor(path: Path?, clazz: Class<T>) :
            this(path, Reflection.createKotlinClass(clazz).createType())

    constructor(addonEnv: AddonEnv, relativePath: String, clazz: Class<T>) : this(
        addonEnv,
        relativePath,
        Reflection.createKotlinClass(clazz).createType()
    )

    constructor(inputStream: InputStream, kType: KType) : this(null, kType) {
        entity = this.getTomlMapper().decode(kType, TomlValue.from(inputStream))
    }

    constructor(inputStream: InputStream, clazz: Class<T>) : this(
        inputStream, Reflection.createKotlinClass(clazz).createType()
    )


    open fun getTomlMapper(): TomlMapper = defaultTomlMapper()
}